### PR TITLE
DashboardSettings: Migrates annotations list & edit view from angular to react and new forms styles

### PR DESCRIPTION
**Pull request by torkelo**
* [x] React list view
* [x] Create new annotation query using same method like variables (ie skip new + add action and automatically create and add new one) 
* [x] Automatically update the model on changes (IE no update button, the Update button feels bad, if you miss to click it your changes are lost) 
* [x] Loading query angular editors
* [x] Load standard react editors 
* [x] Make updating links and annotation arrays immutable so the events and other hacks around updateSubmenuVisibility is no longer needed 
* [x] Refactor out shared components between Links & Annotation list & header  

Todo
* [x] Some ux issues (New button placement, & back button)
* [x] Add tests 
* [x] Align this auto edit & update with how it works in links list view 
* [x] Remove submenuVisibilityChanged it should not be needed  

Screenshot from 2021-03-12 17-40-19

### CODE REVIEW

File: packages/grafana-data/src/types/annotations.ts
Comment (Clarity-89) [inline packages/grafana-data/src/types/annotations.ts:27]:
```suggestion
  [key: string]: any;
```
Nit: More descriptive name.

File: public/app/features/dashboard/components/AnnotationSettings/AngularEditorLoader.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AngularEditorLoader.tsx:21]: Is the `ref` dependency necessary here?

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:28]:
```suggestion
  const { value: ds } = useAsync(() => {
    return getDataSourceSrv().get(annotation.datasource);
  }, [annotation.datasource]);
```

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:31]: Not sure this copy is needed, the copy will be shallow anyway since the `list` contains objects.

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:32]: Maybe better add a check that `editIdx` isn't null?

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:42]: I think this could be handled by `onChange` instead.

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:72]:
```suggestion
          name="name"
          id="name"
```
Good idea to add `id` so the label is properly associated. It also helps later in the tests as it allows matching by role with a `name` parameter.

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:73]: Don't think there's need for `aria-label` if you use id + label combination.

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx:54]:
```suggestion
```

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx:107]:
```suggestion
```

File: public/app/features/dashboard/components/DashboardSettings/AnnotationsSettings.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/DashboardSettings/AnnotationsSettings.tsx:19]: Directly modifying props like this might cause some weird side effects, mostly that components depending on this prop won't detect the change as `dashboard.annotations.list` identity stays the same.

File: public/app/features/dashboard/components/LinksSettings/LinkSettingsEdit.tsx
Comment (Clarity-89) [inline public/app/features/dashboard/components/LinksSettings/LinkSettingsEdit.tsx:87]:
```suggestion
        <Input name="title" id="title" value={linkSettings.title} onChange={onChange} autoFocus={isNew} />
```

**Review by Clarity-89 — COMMENTED** [review]
Good work! 
A few general comments: 
- Noticed that it shows Edit title even when creating a new annotation. 
- Not a big fan of how the new form components look here, especially with checkboxes. 
- Not sure why we need to modify `dashboard` prop directly in components.

File: public/app/features/dashboard/components/AnnotationSettings/AngularEditorLoader.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/AnnotationSettings/AngularEditorLoader.tsx:21]: yea, good catch. I used ref in this one before but then moved it out to a different useEffect

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:28]: nice

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:31]: this is definitely required, and we do not need to do a deep clone here, just the item we are updating, 
The reason we need a new array is otherwise React components that take in the annotation array won't detect the change

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:32]: corrected the type, it cannot be null here

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsEdit.tsx:73]: yea, not sure why that is there, was copied from the LinksEdit page

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx:54]: these are required, cannot return two elements otherwise

File: public/app/features/dashboard/components/DashboardSettings/AnnotationsSettings.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/DashboardSettings/AnnotationsSettings.tsx:19]: yes, this should be done immutably

File: public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx
Comment (Clarity-89) [reply public/app/features/dashboard/components/AnnotationSettings/AnnotationSettingsList.tsx:54]: True, missed that there is a button component at the bottom.

**Review by jackw — APPROVED** [review]
Lovely stuff @torkelo! ⚡

File: public/app/features/dashboard/components/DashboardSettings/DashboardSettingsHeader.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/DashboardSettings/DashboardSettingsHeader.tsx:20]: yes, but this flow we are creating the new annotation query directly and adding it so there is no real new flow. We could maybe show "New" here somehow still

File: public/app/features/dashboard/components/LinksSettings/LinkSettingsList.tsx
Comment (torkelo) [reply public/app/features/dashboard/components/LinksSettings/LinkSettingsList.tsx:52]: yes, been thinking the same, either use the Card component or new table component

File: packages/grafana-data/src/utils/arrayUtils.ts
Comment (dprokop) [reply packages/grafana-data/src/utils/arrayUtils.ts:1]: @jackw add docs annotation and your docs linter goes down by 1 :)

Comment (dprokop) [reply packages/grafana-data/src/utils/arrayUtils.ts:1]: or maybe not, since it's not exported... Correct me if I'm wrong @mckn ?

Comment (jackw) [reply packages/grafana-data/src/utils/arrayUtils.ts]: Spoke with @mckn about these errors and it seems annotating an entire namespace is problematic with docs warnings/errors. Would appreciate a second review before merging.

**Pull request merged**
